fms_extras/utils/generation.py

```diff
@@ -0,0 +1,141 @@
+import functools
+import time
+from typing import Any, Callable, List, MutableMapping, Optional, Union
+
+import torch
+import torch.nn.functional as F
+
+from fms_extras.utils.cache.paged import PagedKVCacheManager
+
+
+def paged_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids: torch.Tensor,
+    kv_cache_manager: PagedKVCacheManager,
+    max_seq_len: int = 2048,
+    max_new_tokens: int = 256,
+    temperature: float = 1.0,
+    top_k: int = 10,
+    do_sample: bool = True,
+    decode_model: Optional[Union[Callable, torch.nn.Module]] = None,
+    # todo: This is a WIP to enable cudagraphs, currently its only for batch_size=1
+    cudagraphs: bool = False,
+):
+    """
+    A trivial generate function that can be used for validation/testing generation using paged attention
+
+    Args:
+        model: Callable or nn.Module
+            A function or nn.Module that takes a batch of input_ids and returns logits
+        kv_cache_manager: PagedKVCacheManager
+            the paged KVCacheManager that handles management of the kv-cache for paged attention
+        max_seq_len: int
+            the max sequence length of the model
+        max_new_tokens: int
+            max tokens to generate
+        temperature: float
+            temperature of softmax when sampling
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+
+    Returns:
+    Tuple[torch.Tensor, int, int]
+        the resulting output tokens, the number of new tokens generated, and the time it took per token in seconds
+    """
+    if decode_model is None:
+        decode_model = model
+
+    bsize = len(input_ids)
+
+    if cudagraphs and bsize != 1:
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1"
+        )
+
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids])
+    n_pads_init = [max_len - seq.size(0) for seq in input_ids]
+    input_ids = torch.stack(
+        [F.pad(input_ids[i], (n_pads_init[i], 0)) for i in range(bsize)]
+    )
+
+    result = input_ids
+    next_input = input_ids
+    kwargs: MutableMapping[str, Any] = dict()
+    kwargs["cache_data"] = None
+    kwargs["use_cache"] = True
+    sequence_ids: Optional[List[int]] = None
+    block_mapping_max = ((max_len + max_new_tokens) // 16) + 1
+    for i in range(max_new_tokens):
+        input_ids = next_input[:, -max_seq_len:]
+
+        # compute the mask
+        if i == 0:
+            is_pad = input_ids == 0
+            mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+            kwargs["mask"] = mask.tril(diagonal=0)
+        else:
+            if i == 1:
+                start_time = time.time()
```

Comment:
shouldn't we include the first pass in the timer?

Reply:
we could, though we were only timing the decode step. Let me check and get back to you how we would like to time.

Reply:
Added timing for the first token and the subsequent decode tokens in latest commit